### PR TITLE
[test] Modernise "Hello, World" tests

### PR DESCRIPTION
**Pull request by victorusu**

### CODE REVIEW

File: cscs-checks/prgenv/helloworld.py [hunk 1]
```diff
@@ -106,123 +119,123 @@ def compile_timer_end(self):
...
-        self.num_tasks_per_node = 1
-        self.num_cpus_per_task = 1
         if (self.current_system.name in ['arolla', 'tsa'] and
-            linkage == 'dynamic'):
+            linking == 'dynamic'):
             self.valid_prog_environs += ['PrgEnv-pgi-nompi',
-                                         'PrgEnv-pgi-nompi-nocuda',
-                                         'PrgEnv-gnu-nompi',
-                                         'PrgEnv-gnu-nompi-nocuda']
+                                        'PrgEnv-pgi-nompi-nocuda',
```
Comment (teojgo) [inline cscs-checks/prgenv/helloworld.py:137]: Align this, one space in.

File: cscs-checks/prgenv/helloworld.py [hunk 2]
```diff
@@ -11,34 +11,43 @@
...
-            self.exclusive_access = True
+    @run_after('init')
+    def set_craylinktype_env_variable(self):
+        self.variables['CRAYPE_LINK_TYPE'] = self.linking
 
-        self.compilation_time_seconds = None
-
-        result = sn.findall(r'Hello World from thread \s*(\d+) out '
+    @run_after('init')
+    def set_sanity_patterns(self):
```
Comment (teojgo) [inline cscs-checks/prgenv/helloworld.py:49]: Use the `@sanity_pattern` decorator here directly.

**Review by teojgo — CHANGES_REQUESTED** [review]

**Comment by victorusu** [comment]
@jenkins-cscs retry all

File: cscs-checks/prgenv/helloworld.py [hunk 3]
```diff
@@ -75,6 +83,9 @@ def num_ranks(match):
                 ),
             )
         )
+
+    @run_after('init')
```
Comment (jjotero) [inline cscs-checks/prgenv/helloworld.py:87]: I think we could try and get the hooks in pipeline-order when possible.
```suggestion
    @run_before('performance')
```

File: cscs-checks/prgenv/helloworld.py [hunk 4]
```diff
@@ -106,123 +118,123 @@ def compile_timer_end(self):
...
 
+    @run_after('init')
+    def update_description(self):
+        self.descr += ' OpenMP ' + self.linking.capitalize()
+
+    @run_after('init')
+    def update_sourcepath(self):
+        self.sourcepath += '_openmp.' + self.lang
+
+    @run_after('init')
```
Comment (jjotero) [inline cscs-checks/prgenv/helloworld.py:184]:
```suggestion
    @run_before('run')
```

File: cscs-checks/prgenv/helloworld.py [hunk 5]
```diff
@@ -106,123 +118,123 @@ def compile_timer_end(self):
...
             self.valid_prog_environs += ['PrgEnv-pgi-nompi',
                                          'PrgEnv-pgi-nompi-nocuda',
                                          'PrgEnv-gnu-nompi',
                                          'PrgEnv-gnu-nompi-nocuda']
 
+    @run_after('init')
+    def update_description(self):
+        self.descr += ' OpenMP ' + self.linking.capitalize()
+
+    @run_after('init')
```
Comment (jjotero) [inline cscs-checks/prgenv/helloworld.py:180]:
```suggestion
    @run_before('compile')
```

File: cscs-checks/prgenv/helloworld.py [hunk 6]
```diff
@@ -106,123 +118,123 @@ def compile_timer_end(self):
...
             self.valid_prog_environs += ['PrgEnv-pgi-nompi',
                                          'PrgEnv-pgi-nompi-nocuda',
                                          'PrgEnv-gnu-nompi',
                                          'PrgEnv-gnu-nompi-nocuda']
 
+    @run_after('init')
+    def update_description(self):
+        self.descr += ' Serial ' + self.linking.capitalize()
+
+    @run_after('init')
```
Comment (jjotero) [inline cscs-checks/prgenv/helloworld.py:144]:
```suggestion
    @run_before('compile')
```

File: cscs-checks/prgenv/helloworld.py [hunk 7]
```diff
@@ -106,123 +118,123 @@ def compile_timer_end(self):
...
+    def update_description(self):
+        self.descr += ' MPI ' + self.linking.capitalize()
 
-        # for the MPI test the self.num_tasks_per_node should always be one. If
-        # not, the test will fail for the total number of lines in the output
-        # file is different then self.num_tasks * self.num_tasks_per_node
-        self.num_tasks = 2
-        self.num_tasks_per_node = 1
-        self.num_cpus_per_task = 1
+    @run_after('init')
```
Comment (jjotero) [inline cscs-checks/prgenv/helloworld.py:205]:
```suggestion
    @run_before('compile')
```

File: cscs-checks/prgenv/helloworld.py [hunk 8]
```diff
@@ -106,123 +118,123 @@ def compile_timer_end(self):
...
         }
 
-        self.num_tasks = 6
-        self.num_tasks_per_node = 3
-        self.num_cpus_per_task = 4
+    @run_after('init')
+    def update_description(self):
+        self.descr += ' MPI + OpenMP ' + self.linking.capitalize()
 
+    @run_after('init')
```
Comment (jjotero) [inline cscs-checks/prgenv/helloworld.py:232]:
```suggestion
    @run_before('compile')
```

File: cscs-checks/prgenv/helloworld.py [hunk 9]
```diff
@@ -106,123 +118,123 @@ def compile_timer_end(self):
...
+    def update_sourcepath(self):
+        self.sourcepath += '_mpi_openmp.' + self.lang
+
+    @run_after('init')
```
Comment (jjotero) [inline cscs-checks/prgenv/helloworld.py:236]:
```suggestion
    @run_before('run')
```

**Review by jjotero — COMMENTED** [review]
My comments are only related to the actual stages where hooks should be attached to. I think that as a general rule (we should probably have a rationale doc covering this) whichever test attribute cannot be set directly in the class body, should be placed in a hook that executes just in time for what the hook should do. Otherwise we might end up with loads of `@run_after('init')` hooks, which kind of defeats the purpose of having hooks to begin with. It also gets a bit hard to see which of those hooks are actually needed after init, and which others could actually execute at a later stage.

**Comment by teojgo** [comment]
I will address the PR comments on this.

**Review by jjotero — APPROVED** [review]

File: cscs-checks/prgenv/helloworld.py [hunk 10]
```diff
@@ -11,34 +11,43 @@
...
+    sourcepath = 'hello_world'
+    build_system = 'SingleSource'
+    valid_systems = ['daint:gpu', 'daint:mc', 'dom:gpu', 'dom:mc',
+                     'arolla:cn', 'arolla:pn', 'tsa:cn', 'tsa:pn']
+    valid_prog_environs = ['PrgEnv-aocc', 'PrgEnv-cray',
+                           'PrgEnv-cray_classic', 'PrgEnv-gnu',
+                           'PrgEnv-intel', 'PrgEnv-pgi',
+                           'PrgEnv-gnu-nocuda', 'PrgEnv-pgi-nocuda']
+    exclusive_access = True
+    compilation_time_seconds = None
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:31]: Shouldn't that be a variable and not a class variable? Each test is setting that.

File: cscs-checks/prgenv/helloworld.py [hunk 11]
```diff
@@ -11,34 +11,43 @@
...
-                              'arolla:cn', 'arolla:pn', 'tsa:cn', 'tsa:pn']
-        if linkage == 'dynamic':
+    linking = parameter(['dynamic', 'static'])
+    lang = parameter(['c', 'cpp', 'f90'])
+    prgenv_flags = {}
+    lang_names = {
+        'c': 'C',
+        'cpp': 'C++',
+        'f90': 'Fortran 90'
+    }
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:21]: This is very local to the `set_description` hook, so I'd move it there.

File: cscs-checks/prgenv/helloworld.py [hunk 12]
```diff
@@ -11,34 +11,43 @@
...
+    compilation_time_seconds = None
+    maintainers = ['VH', 'EK']
+    tags = {'production', 'craype'}
+
+    @run_after('init')
+    def set_description(self):
+        self.descr = f'{self.lang_names[self.lang]} Hello, World'
+
+    @run_after('init')
+    def add_dynamic_only_systems(self):
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:40]:
```suggestion
    def adapt_valid_systems(self):
```

File: cscs-checks/prgenv/helloworld.py [hunk 13]
```diff
@@ -11,34 +11,43 @@
...
-                                    'PrgEnv-cray_classic', 'PrgEnv-gnu',
-                                    'PrgEnv-intel', 'PrgEnv-pgi',
-                                    'PrgEnv-gnu-nocuda', 'PrgEnv-pgi-nocuda']
-
-        if self.current_system.name in ['arolla', 'tsa']:
-            self.exclusive_access = True
-
-        self.compilation_time_seconds = None
+    @run_before('compile')
+    def set_craylinktype_env_variable(self):
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:45]:
```suggestion
    def setup_linking_type(self):
```

File: cscs-checks/prgenv/helloworld.py [hunk 14]
```diff
@@ -11,34 +11,43 @@
...
+        self.variables['CRAYPE_LINK_TYPE'] = self.linking
 
-        result = sn.findall(r'Hello World from thread \s*(\d+) out '
+    @sanity_function
+    def set_sanity_patterns(self):
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:49]:
```suggestion
    def assert_hello_world(self):
```

File: cscs-checks/prgenv/helloworld.py [hunk 15]
```diff
@@ -75,6 +83,9 @@ def num_ranks(match):
                 ),
             )
         )
+
+    @run_before('performance')
+    def set_performance_patterns(self):
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:88]:
```suggestion
    def capture_build_time(self):
```

File: cscs-checks/prgenv/helloworld.py [hunk 16]
```diff
@@ -84,14 +95,15 @@ def num_ranks(match):
             }
         }
 
-        self.maintainers = ['VH', 'EK']
-        self.tags = {'production', 'craype'}
-
     @run_before('compile')
     def setflags(self):
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:99]: I'd suggest something like this and you could merge it with the other hook setting the linking type.

```suggestion
    def prepare_build(self):
```

File: cscs-checks/prgenv/helloworld.py [hunk 17]
```diff
@@ -106,123 +118,123 @@ def compile_timer_end(self):
...
                                          'PrgEnv-gnu-nompi',
                                          'PrgEnv-gnu-nompi-nocuda']
 
+    @run_after('init')
+    def update_description(self):
+        self.descr += ' Serial ' + self.linking.capitalize()
+
+    @run_before('compile')
+    def update_sourcepath(self):
+        self.sourcepath += '_serial.' + self.lang
```
Comment (vkarak) [inline cscs-checks/prgenv/helloworld.py:146]:
```suggestion
        self.sourcepath += f'_serial.{self.lang}'
```

**Review by vkarak — CHANGES_REQUESTED** [review]

**Review by vkarak — APPROVED** [review]
lgtm

**Review by teojgo — APPROVED** [review]

**Pull request merged**